README
This is a "plugin" for the Video Disk Recorder (VDR).

Written by:                  Dominic Morris <dom@suborbital.org.uk>

Project's homepage:          http://www.rst38.org.uk/vdr/decruft

Latest version available at: http://www.rst38.org.uk/vdr/decruft

See the file COPYING for license information.

Description:

Clean the cruft from your channels file - i.e. turn on add channels without
fear of being overwhelmed.

Sort your channels into logical groups.

Example config file is in examples/decruft.conf

Apply the patch in patches before compilation - if you've already got the
source dependent channel insertion patch then you won't need this.

TODO:

- Cleanup on demand (done 0.0.4)
- Configure cleanup period (it's 10 mins ATM) (will never be done, periodic
cleanup was removed in 0.0.4)
- Configure filters using OSD?


HISTORY
VDR Plugin 'decruft' Revision History
-------------------------------------

2005-03-11: Version 0.0.4

- Decrufting is only available by the main menu - the periodic cleaning was
causing too many problems with other plugins

2005-02-20: Version 0.0.3

- Add in support for freq=,pol= for blacklisting a transponder

2005-02-15: Version 0.0.2

- Fix some problems if source,provider or name were specified as a filter -
the remainder of the line was ignored
- Add grouping of channels

2005-02-14: Version 0.0.1

- Initial revision.


cruft.c
/*
 * decruft: A plugin for the Video Disk Recorder
 *
 * See the README file for copyright information and how to reach the author.
 *
 * $Id: cruft.c,v 1.4 2005/02/20 18:34:01 dom Exp $
 *
 * Purge or move channels based on a spec
 *
 * Moving/grouping based on the Source dependent channel insertion patch
 *
 * TODO: Reloading of config
 * TODO: Better file format?
 */


#include <vdr/channels.h>
#include <vdr/dvbdevice.h>

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <ctype.h>
#include <regex.h>

typedef struct {
    char     *group_name;
    int       num_cas;
    int      *cas;
    int       num_vpids;
    int      *vpids;
    int       num_apids;
    int      *apids;
    int       num_dpids;
    int      *dpids;
    int       num_tpids;
    int      *tpids;
    int       num_provs;
    char    **provs;
    regex_t **provs_regex;
    int       num_names;
    char    **names;
    regex_t **names_regex;
    int       num_freqs;
    int      *freqs;
    int       num_pols;
    char     *pols;
    int       num_sources;
    char    **sources;
} setting_t;


static int          num_clean = 0;
static setting_t  **clean     = NULL;
static int          num_keep  = 0;
static setting_t  **keep      = NULL;
static int          num_groups = 0;
static setting_t  **groups     = NULL;

/* Cleanup a settings_t */
void free_settings(setting_t *s)
{
    int   i;
    if ( s->cas ) 
        free(s->cas);
    if ( s->vpids )
        free(s->vpids);
    if ( s->apids )
        free(s->apids);
    if ( s->dpids )
        free(s->dpids);
    if ( s->tpids )
        free(s->tpids);
    for ( i = 0; i < s->num_provs; i++ ) {
        free(s->provs[i]);
        if ( s->provs_regex[i] ) {
                regfree(s->provs_regex[i]);
                free(s->provs_regex[i]);
        }
    }
    if ( s->provs ) {
        free(s->provs);
        free(s->provs_regex);
    }
    for ( i = 0; i < s->num_names; i++ ) {
        free(s->names[i]);
        if ( s->names_regex[i] ) {
            if ( s->names_regex[i] ) {
                regfree(s->names_regex[i]);
                free(s->names_regex[i]);
            }
        }
    }
    if ( s->names ) {
        free(s->names);
        free(s->names_regex);
    }
    for ( i = 0; i < s->num_sources; i++ ) {
        free(s->sources[i]);
    }
    if ( s->sources ) {
        free(s->sources);
    }
    if ( s->freqs ) {
        free(s->freqs);
    }
    if ( s->pols ) {
        free(s->pols);
    }
    if ( s->group_name ) {
        free(s->group_name);
    }
    free(s);
}


/* TODO: This parsing code (and file format) sucks, it really needs to
 *  be fixed, but...I guess it does the job...
 */
static char *check_arg(char *ptr, char *match)
{
    while ( isspace(*ptr) )
        ptr++;
    if ( strncmp(ptr,match,strlen(match)) ) {
        return NULL;
    }
    ptr += strlen(match);
    while ( isspace(*ptr) )
        ptr++;
    return ptr;
}

static void read_ints(char *line, int *count, int **dest, bool hex=false)
{
    char   tempbuf[10];
    int    num;
    int    pos = 0;

    while (*line && *line != ';' ) {
        if (*line == ',' || !isxdigit(*line) ) {
            tempbuf[pos] = 0;
            if ( pos == 0 ) {
                line++;
                continue;
            }
            if ( sscanf(tempbuf,"%x",&num) == 1 ) {
                *dest = (int *)realloc(*dest, (*count + 1) * sizeof(int) );
                (*dest)[*count] = num;
                (*count)++;
            }
            line++;
            pos = 0;
            continue;
        } else {
            tempbuf[pos++] = *line;
            line++;
        }
    }
    /* Pick up the last one */
    if ( pos != 0 ) {
	const char *fmt;
        tempbuf[pos] = 0;
	if ( hex ) {
	    fmt = "%x";
	} else {
	    fmt = "%d";
	}
        if ( sscanf(tempbuf,fmt,&num) == 1 ) {
            *dest = (int *)realloc(*dest, (*count + 1) * sizeof(int) );
            (*dest)[*count] = num;
            (*count)++;
        }
    }
}

static int parse_line(setting_t *settings, char *line)
{
    char   *ptr = line;
    char   *temp;
    char   *end;
    int     i;


    while ( isspace(*ptr) )
        ptr++;

    while ( *ptr ) {
        if ( ( temp = check_arg(ptr,"ca=") ) != NULL ) {
            read_ints(temp,&settings->num_cas,&settings->cas,true);
        } else if ( (temp = check_arg(ptr,"provider=") ) != NULL ) {
            if ( (end = strchr(temp,';')) || (end=strchr(temp,'\n') ) ) {
                ptr = end + 1;
                *end = 0;
                i = settings->num_provs++;
                settings->provs = (char **)realloc(settings->provs, settings->num_provs * sizeof(char *));
                settings->provs_regex = (regex_t **)realloc(settings->provs_regex, settings->num_provs * sizeof(char *));
                settings->provs[i] = strdup(temp);
                /* Compile the regex */
                settings->provs_regex[i] = (regex_t *)calloc(1,sizeof(regex_t));
                if ( regcomp(settings->provs_regex[i], temp, REG_EXTENDED) != 0 ) {
                    free(settings->provs_regex[i]);
                    settings->provs_regex[i] = NULL;
                }
		temp = ptr;
            } else {
                /* Hmmmm, fail gracefully */
            }
        } else if ( (temp = check_arg(ptr,"name=") ) != NULL ) {
            if ( (end = strchr(temp,';')) || (end=strchr(temp,'\n') ) ) {
                ptr = end + 1;
                *end = 0;
                i = settings->num_names++;
                settings->names = (char **)realloc(settings->names, settings->num_names * sizeof(char *));
                settings->names_regex = (regex_t **)realloc(settings->names_regex, settings->num_names * sizeof(char *));
                settings->names[i] = strdup(temp);
                /* Compile the regex */
                settings->names_regex[i] = (regex_t *)calloc(1,sizeof(regex_t));
                if ( regcomp(settings->names_regex[i], temp, REG_EXTENDED) != 0 ) {
                    free(settings->names_regex[i]);
                    settings->names_regex[i] = NULL;
                }
		temp = ptr;
            } else {
                /* Hmmmm, fail gracefully */
            }
        } else if ( (temp = check_arg(ptr,"source=") ) != NULL ) {
            if ( (end = strchr(temp,';')) || (end=strchr(temp,'\n') ) ) {
                ptr = end + 1;
                *end = 0;
                i = settings->num_sources++;
                settings->sources = (char **)realloc(settings->sources, settings->num_sources * sizeof(char *));
                settings->sources[i] = strdup(temp);
		temp = ptr;
            } else {
                /* Hmmmm, fail gracefully */
            }
        } else if ( (temp = check_arg(ptr,"vpid=") ) != NULL ) {
            read_ints(temp,&settings->num_vpids,&settings->vpids);
        } else if ( (temp = check_arg(ptr,"apid=") ) != NULL ) {
            read_ints(temp,&settings->num_apids,&settings->apids);
        } else if ( (temp = check_arg(ptr,"dpid=") ) != NULL ) {
            read_ints(temp,&settings->num_apids,&settings->apids);
        } else if ( (temp = check_arg(ptr,"tpid=") ) != NULL ) {
            read_ints(temp,&settings->num_dpids,&settings->dpids);
        } else if ( (temp = check_arg(ptr,"group=") ) != NULL ) {
            if ( (end = strchr(temp,';')) || (end=strchr(temp,'\n') ) ) {
                ptr = end + 1;
                *end = 0;
                if ( settings->group_name )
                    free(settings->group_name);
                settings->group_name = strdup(temp);
	        temp = ptr;
            } else {
                /* Hmmm fail gracefully */
            }
        } else if ( (temp = check_arg(ptr,"freq=") ) != NULL ) {
            read_ints(temp,&settings->num_freqs,&settings->freqs);
        } else if ( (temp = check_arg(ptr,"pol=") ) != NULL ) {
            if ( (end = strchr(temp,';')) || (end=strchr(temp,'\n') ) ) {
                ptr = end + 1;
                *end = 0;
                i = settings->num_pols++;
                settings->pols = (char *)realloc(settings->pols, settings->num_pols);
                settings->pols[i] = tolower(*temp);
            } else {
                /* Hmmmm Fail gracefully */
            }
        } else {
            temp = ptr + 1;
        }
        ptr = temp;
    }
    return 1;
}


int parse_file(const char *filename)
{
    char          buf[1024];
    FILE         *fp;
    setting_t    *setting;
    char         *ptr;
    int           i;

    if ( ( fp = fopen(filename,"r")) == NULL ) {
        return -1;
    }

    while ( fgets(buf,sizeof(buf),fp) != NULL ) {
        if ( buf[0] == '#' || buf[0] == ';' ) {
            continue;
        }
        if ( strncmp(buf,"clean:",strlen("clean:")) == 0) {
            ptr = buf + strlen("clean:") + 1;
            setting = (setting_t *)calloc(1,sizeof(*setting));
            if ( parse_line(setting,ptr) ) {
                i = num_clean++;
                clean = (setting_t **)realloc(clean, num_clean * sizeof(setting_t *));
                clean[i] = setting;
            } else {
                free(setting);
            }
        } else if ( strncmp(buf,"keep:",strlen("keep:")) == 0 ) {
            ptr = buf + strlen("keep:") + 1;
            setting = (setting_t *)calloc(1,sizeof(*setting));
            if ( parse_line(setting,ptr) ) {
                i = num_keep++;
                keep = (setting_t **)realloc(keep, num_keep * sizeof(setting_t *));
                keep[i] = setting;
            } else {
                free(setting);
            }
        } else if ( strncmp(buf,"group:",strlen("group:")) == 0 ) {
            ptr = buf + strlen("group:") + 1;
            setting = (setting_t *)calloc(1,sizeof(*setting));
            if ( parse_line(setting,ptr) ) {
                i = num_groups++;
                groups = (setting_t **)realloc(groups, num_groups * sizeof(setting_t *));
                groups[i] = setting;
            } else {
                free(setting);
	    }
        }
    }
    return 1;
}

/** \brief Check a channel against an individual settings
 *
 *  \param channel
 *  \param setting
 *
 *  \retval true - matched
 *  \retval false - not matched
 */
static bool CheckSettings(cChannel *Channel, setting_t *settings)
{
    int     tests = 0;
    int     match = 0;
    int     i,j;
    /* Check CAs */

               cDvbTransponderParameters dtp(Channel->Parameters());
    if ( settings->num_cas ) {
        tests++;
        for ( j = 0; j < settings->num_cas; j++ ) {
            for ( i = 0; i < MAXCAIDS; i++ ) {
                int ca = Channel->Ca();
                if ( ca == settings->cas[j] ) {
                    match++;
                    goto exitca;
                }
            }
        }
    exitca:
        ;
    }
    if ( settings->num_vpids ) {
        tests++;
        for ( i = 0; i < settings->num_vpids; i++ ) {
            if ( Channel->Vpid() == settings->vpids[i] ) {
                match++;
                break;
            }
        }
    }
    if ( settings->num_apids ) {
        tests++;
        /* We only really want to check for 0 as the first pid */
        for ( j = 0; j < settings->num_apids; j++ ) {
            if ( settings->apids[j] == Channel->Apid(0) ) {
                match++;
                break;
            }
        }
    }
    if ( settings->num_dpids ) {
        tests++;
        /* We only really want to check for 0 as the first pid */
        for ( j = 0; j < settings->num_dpids; j++ ) {
            if ( settings->dpids[j] == Channel->Apid(0) ) {
                match++;
                break;
            }
        }
    }
    if ( settings->num_tpids ) {
        tests++;
        /* We only really want to check for 0 as the first pid */
        for ( j = 0; j < settings->num_tpids; j++ ) {
            if ( settings->tpids[j] == Channel->Tpid() ) {
                match++;
                break;
            }
        }
    }

    if ( settings->num_provs ) {
        tests++;
        for ( j = 0; j < settings->num_provs; j++ ) {
            if ( settings->provs_regex[j] ) {
                if ( regexec(settings->provs_regex[j],Channel->Provider(),0,NULL,0) == 0 ) {
                    match++;
                    break;
                }
            }
        }
    }

   if ( settings->num_names ) {
        tests++;
        for ( j = 0; j < settings->num_names; j++ ) {
            if ( settings->names_regex[j] ) {
                if ( regexec(settings->names_regex[j],Channel->Name(),0,NULL,0) == 0 ) {
                    match++;
                    break;
                }
            }
        }
    }

   if ( settings->num_sources ) {
       char *source = strdup(*cSource::ToString(Channel->Source()));
       tests++;
       for ( j = 0; j < settings->num_sources; j++ ) {
           if ( strcmp(source,settings->sources[j]) == 0 ) {
               match++;
               break;
           }
       }
       free(source);
   }
   if ( settings->num_freqs ) {
       int  freq = Channel->Frequency();
       tests++;
       for ( j = 0; j < settings->num_freqs; j++ ) {
           if ( freq == settings->freqs[j] ) {
               match++;
               break;
           }
       }
   }
   if ( settings->num_pols ) {
       char  pol = tolower(dtp.Polarization());
       tests++;
       for ( j = 0; j < settings->num_pols; j++ ) {
           if ( settings->pols[j] == pol ) {
               match++;
               break;
           }
       }
   }
   if ( tests == match && tests != 0 ) {
       return true;
   }
   return false;
}

/** \brief See whether a channel should be kept or not
 *
 *  \param channel The incoming channel
 *
 *  \retval false - The channel should be removed
 *  \retval true  - The channel should be retained
 */
bool CheckChannel(cChannel *channel)
{
    bool          res = true;  /* Default to keeping it (be safe) */
    int           i;
	
    printf("Checking channel <%s> Source <%s>\n",channel->Name(), *cSource::ToString(channel->Source()));

    for ( i = 0; i < num_clean; i++ ) {
        if ( CheckSettings(channel,clean[i]) ) {
            res = false;
            break;
        }
    }

    if ( res == false ) {
        for ( i = 0; i < num_keep; i++ ) {
            if ( CheckSettings(channel,keep[i] ) ) {
                res = true;
                break;
            }
        }
    }
    return res;
}

/** \brief Get a channel by name
 *
 *  \param name The name to find
 *
 *  \retval NULL - not found
 *  \retval other - The cChannel
 */
static cChannel *ChannelGetByName(char *name)
{
    for (cChannel *channel = Channels.First(); channel; channel = Channels.Next(channel)) {
        if (strcmp(channel->Name(), name)==0)
            return channel;
    }
    return NULL;
}


/** \brief Try and move a channel based on a settings group 
 *
 *  \param channel The channel to test
 *  \param settings The settings to test against
 *
 *  \retval true - Channel moved
 *  \retval false - Channel not moved
 */
static bool CheckChannelMoveReal(cChannel *channel, setting_t *settings)
{
    if ( CheckSettings(channel,settings) ) {
        /* If it matches the filter then we should attempt to move it */
        cChannel *groupSep;
        int       idx;

        /* Get the current group separator */
        idx = Channels.GetPrevGroup(channel->Index());

        if ( idx != -1 ) {
            groupSep = Channels.Get(idx);
            if ( strcmp(groupSep->Name(),settings->group_name) == 0 ) {
                /* Already in the group so no need to move */
                return false;
            }
        }
        
        /* Find which group its meant to be in */
        groupSep = ChannelGetByName(settings->group_name);
            
        /* Not defined, so define it */
        if ( !groupSep ) {
            groupSep = new cChannel();
            char groupSepString[512];
            sprintf(groupSepString, ":%s", settings->group_name);
            groupSep->Parse(groupSepString);
            Channels.Add(groupSep);
        }
        /* Move the channel to the end of the group */
        if ( Channels.GetNextGroup(groupSep->Index()) >= 0 ) {
            groupSep = Channels.Get(Channels.GetNextGroup(groupSep->Index()));
        } else {
            groupSep = Channels.Last();
        }
        esyslog("Moving channel <%s;%s> to group <%s>\n",channel->Name(),channel->Provider(), settings->group_name);
        Channels.Move(channel,groupSep);
        return true;
    }
    return false;
}

/** \brief Group channels as appropriate
 *
 *  \param channel The channel to organise
 *
 *  \retval true - Moved channel
 *  \retval false - Didn't move channel
 */
bool CheckChannelMove(cChannel *channel)
{
    bool  ret = false;
    int   i;

    printf("Check if <%s> needs to be moved\n",channel->Name());

    for ( i = 0; i < num_keep; i++ ) {
        if ( keep[i]->group_name != NULL ) {
            if ( CheckChannelMoveReal(channel,keep[i]) == true ) {
                ret = true;
	    }
        }
    }
    for ( i = 0; i < num_groups; i++ ) {
        if ( groups[i]->group_name != NULL ) {
            if ( CheckChannelMoveReal(channel,groups[i]) == true ) {
                ret = true;
	    }
        }
    }
    return ret;
}


cleanthread.c
/*
 * decruft: A plugin for the Video Disk Recorder
 *
 * See the README file for copyright information and how to reach the author.
 *
 * $Id: cleanthread.c,v 1.3 2005/03/11 18:49:11 dom Exp $
 *
 * Main thread - based on the epgsearch timer thread code
 *
 * TODO: Trigger a rescan 
 */


#include <vdr/channels.h>
#include <vdr/timers.h>
#include "cleanthread.h"
#include "cruft.h"


void sleepMSec(long ms)
{
#if VDRVERSNUM >= 10314
    cCondWait::SleepMs(ms);
#else
    usleep(1000 * ms);
#endif
}

void sleepSec(long s)
{
    sleepMSec(s * 1000);
}

static cCruftCleanThread *Instance = NULL;

void startDecruft()
{
    Instance->Trigger();
}



cCruftCleanThread::cCruftCleanThread(void)
    : cThread("Decruft: cleaning")
{
    m_Active = false;
    nextUpdate = -1;
}

cCruftCleanThread::~cCruftCleanThread()
{
    if (m_Active )
        Stop();
}

void cCruftCleanThread::Init(void)
{
    if ( Instance == NULL ) {
        Instance = new cCruftCleanThread;
        Instance->Start();
    }
}

void cCruftCleanThread::Exit(void)
{
    if (Instance != NULL ) {
        Instance->Stop();
        DELETENULL(Instance);
    }
}

void cCruftCleanThread::Stop(void)
{
    m_Active = false;
    Cancel(3);
}


void cCruftCleanThread::Trigger()
{
    nextUpdate = time(NULL);
}
    

void cCruftCleanThread::Action(void)
{
    m_Active = true;

    // let VDR do its startup
    sleepSec(15);

    while (m_Active ) {
        time_t now = time(NULL);
        if (now >= nextUpdate && nextUpdate > 0  ) {
            if ( Channels.BeingEdited() ) {
		printf("Channels are being edited");
                sleepSec(1);
                continue;
            }

            /* Grab the channels lock */
            if (!Channels.Lock(true,10)) {
		printf("Can't lock channels\n");
		sleepSec(1);
                continue;
            }

            for ( cChannel *channel = Channels.First(); channel ; ) {
                cChannel *next = Channels.Next(channel);

                /* If it's not a separator then process it */
                if ( !channel->GroupSep() ) {
                    if ( CheckChannel(channel) == false ) {
                        /* Check for active timers */
                        bool del = true;
                        for (cTimer *ti = Timers.First(); ti; ti = Timers.Next(ti)) {
                            if (ti->Channel() == channel) {
                                del = false;
                            }
                        }
                        if ( del ) {
                            esyslog("Deleting channel <%s;%s>\n",channel->Name(),channel->Provider());
                            Channels.Del(channel);
                            Channels.ReNumber();
                        } else {
                            /* We might hit the same channels again, but the
                               second time they won't be moved...nasty 
                            */
                            CheckChannelMove(channel);
                        }

                    } else {
			CheckChannelMove(channel);
		    }
                }
                channel = next;
            }
            /* And unlock them */
            Channels.SetModified(true);
            Channels.Unlock();
            // nextUpdate = m_lastUpdate + 600; /* FIXME: run every 10 minutes */
	    nextUpdate = -1;	// Make this run the once
        }
        sleepSec(5);
    }
    dsyslog("Decruft: Leaving cleaning thread");
}
                


decruft.c
/*
 * decruft.c: A plugin for the Video Disk Recorder
 *
 * See the README file for copyright information and how to reach the author.
 *
 * $Id: decruft.c,v 1.3 2005/03/11 18:49:11 dom Exp $
 *
 * TODO: Configuration
 * TODO: Option to scan on demand (0.0.4)
 */

#include <vdr/plugin.h>
#include "cleanthread.h"
#include "menu.h"
#include <time.h>
#include "cruft.h"

static const char *VERSION        = "0.0.4";
static const char *DESCRIPTION    = "Remove the cruft from your channels";
static const char *MAINMENUENTRY  = "Decruft";

class cPluginDecruft : public cPlugin {
private:
  // Add any member variables or functions you may need here.
       time_t last_decruft;
public:
  cPluginDecruft(void);
  virtual ~cPluginDecruft();
  virtual const char *Version(void) { return VERSION; }
  virtual const char *Description(void) { return DESCRIPTION; }
  virtual const char *CommandLineHelp(void);
  virtual bool ProcessArgs(int argc, char *argv[]);
  virtual bool Initialize(void);
  virtual bool Start(void);
  virtual void Stop(void);
  virtual void Housekeeping(void);
  virtual const char *MainMenuEntry(void) { return MAINMENUENTRY; }
  virtual cOsdObject *MainMenuAction(void);
  virtual cMenuSetupPage *SetupMenu(void);
  virtual bool SetupParse(const char *Name, const char *Value);
       virtual const char **SVDRPHelpPages(void);
       virtual cString SVDRPCommand(const char *Command, const char *Option, int &ReplyCode);
  };

cPluginDecruft::cPluginDecruft(void)
{
  // Initialize any member variables here.
  // DON'T DO ANYTHING ELSE THAT MAY HAVE SIDE EFFECTS, REQUIRE GLOBAL
  // VDR OBJECTS TO EXIST OR PRODUCE ANY OUTPUT!
       g_DecruftSetup.cleanEvery=12;
}

cPluginDecruft::~cPluginDecruft()
{
  // Clean up after yourself!
}

const char *cPluginDecruft::CommandLineHelp(void)
{
  // Return a string that describes all known command line options.
  return NULL;
}

bool cPluginDecruft::ProcessArgs(int argc, char *argv[])
{
  // Implement command line argument processing here if applicable.
  return true;
}

bool cPluginDecruft::Initialize(void)
{
  // Initialize any background activities the plugin shall perform.
  return true;
}

bool cPluginDecruft::Start(void)
{
  // Start any background activities the plugin shall perform.
  parse_file(AddDirectory(cPlugin::ConfigDirectory(), "decruft.conf"));
  cCruftCleanThread::Init();
  time(&last_decruft);
  return true;
}

void cPluginDecruft::Stop(void)
{
  // Stop any background activities the plugin shall perform.
}

void cPluginDecruft::Housekeeping(void)
{
       time_t now;
       time(&now);
       esyslog("housekeeping decruft %d %d %d",now,last_decruft,now-last_decruft);
       if (now-last_decruft > g_DecruftSetup.cleanEvery * 3600){ // every 12 std
               last_decruft = now;
               startDecruft();
       }

  // Perform any cleanup or other regular tasks.
}

cOsdObject *cPluginDecruft::MainMenuAction(void)
{
  // Perform the action when selected from the main VDR menu.
  startDecruft();
  return NULL;
}

cMenuSetupPage *cPluginDecruft::SetupMenu(void)
{
  // Return a setup menu in case the plugin supports one.
       return new cSetupMenu();
}

bool cPluginDecruft::SetupParse(const char *Name, const char *Value)
{
  // Parse your own setup parameters and store their values.
       if      (!strcasecmp(Name, "CleanEvery"))  g_DecruftSetup.cleanEvery = atoi(Value);
       else return false;
       return true;

}
const char **cPluginDecruft::SVDRPHelpPages(void)
{
    // Return help text for SVDRP commands this plugin implements
   static const char *HelpPage[] =
    {
        "CLEAN\n"
        "     Clean now channels.conf.",
        NULL
    };
    return HelpPage;
    return NULL;
}



cString cPluginDecruft::SVDRPCommand(const char *Command, const char *Option, int &ReplyCode)
{
    // Process SVDRP commands this plugin implements
               if (!strcasecmp(Command,"CLEAN")){
                       startDecruft();
                       return tr("decruft cleaning channels.conf ...");
               }
    return NULL;
}


VDRPLUGINCREATOR(cPluginDecruft); // Don't touch this!
